### PR TITLE
Add binary search

### PR DESCRIPTION
**Pull request by Archargelod**
I used linear_search.nim as base, so `checkBinarySearch` template is redundant, but could be useful in future for other implementations. I am also not sure if it is a good idea or not to check if array is sorted at the start of function; for example:
```
from std/algorithm import isSorted, SortOrder

func binarySearch*[T:Ordinal](arr: openArray[T], key: T): Option[Natural] {.raises: [ValueError].} =
  if not arr.isSorted(Ascending):
    raise newException(ValueError, "Array is not sorted or not in ascending order.")
```

### CODE REVIEW

**Review by SatinWuker — COMMENTED** [review]
The code looks very good! However, I recommend adding a recursive binary search in order to provide another alternative to people who wants to learn binary search.

**Comment by Archargelod** [comment]
> I recommend adding a recursive binary search in order to provide another alternative to people who wants to learn binary search. See https://www.geeksforgeeks.org/binary-search/

Thank you for the link! I added recursive version and fixed a little inconsistency: `mid.Natural`->`Natural(mid)`.

**Review by SatinWuker — APPROVED** [review]
Looks great!

File: searches/binary_search.nim [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 6, some(Natural(5)))
+
+    test "Search in an even-length array with a valid key in the middle":
+      var arr = [0, 1, 2, 3, 5, 6]
+      checkBinarySearch(arr, 3, some(Natural(3)))
+
+    test "Search in an odd-length array with a valid key in the middle":
+      var arr = [0, 1, 2, 3, 5]
+      checkBinarySearch(arr, 2, some(Natural(2)))
```
Comment (ZoomRmc) [inline searches/binary_search.nim:101]:
```suggestion
    const
      empty: array[0, int] = []
      arr = [0, 1, 2, 3, 5, 6]
      odd = [0, 1, 2, 3, 5]
      chars = ['0', '1', '2', '3', '5', 'a']

    test "Empty array":
      checkBinarySearch(empty, 5, none(Natural))

    test "Matching value in an int array":
      checkBinarySearch(arr, 5, some(Natural(4)))

    test "Missing value":
      checkBinarySearch(arr, 4, none(Natural))

    test "Matching value in a char array":
      checkBinarySearch(chars, '5', some(Natural(4)))

    test "Matching key at the start":
      checkBinarySearch(arr, 0, some(Natural(0)))

    test "Matching key at the end":
      checkBinarySearch(arr, 6, some(Natural(5)))

    test "Even-length array with a matching key in the middle":
      checkBinarySearch(arr, 3, some(Natural(3)))

    test "Odd-length array with a matching key in the middle":
      checkBinarySearch(odd, 2, some(Natural(2)))
```

File: searches/binary_search.nim [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    if arr[mid] == key: 
+      return some(Natural(mid))
+
+    if key < arr[mid]: 
+      return binarySearchRecursive(arr, left, mid - 1, key)
+    else:
+      return binarySearchRecursive(arr, mid + 1, right, key)
+
+  return none(Natural)
+
```
Comment (ZoomRmc) [inline searches/binary_search.nim:66]:
```suggestion

func binarySearchRecursive*[T:Ordinal](arr: openArray[T], key: T): Option[Natural] =
  ## Recursive implementation of binary search for an array sorted in ascending order
  binarySearchRecursive(arr, 0, arr.high(), key)

```
It's idiomatic for Nim to use overloads to simplify user-facing API.

File: searches/binary_search.nim [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    if key < arr[mid]: 
+      right = mid - 1
+    else:
+      left = mid + 1
+
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  return none(Natural)
+
+func binarySearchRecursive*[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```
Comment (ZoomRmc) [inline searches/binary_search.nim:52]:
```suggestion
func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```

*If the suggestion of using an overload for a public func below is accepted.

File: searches/binary_search.nim [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      return binarySearchRecursive(arr, mid + 1, right, key)
+
+  return none(Natural)
+
+when isMainModule:
+  import unittest
+
+  template checkBinarySearch[T:Ordinal](arr: openArray[T], key: T, expected: Option[Natural]): untyped =
+    check binarySearchIterative(arr, key) == expected
+    check binarySearchRecursive(arr, arr.low(), arr.high(), key) == expected
```
Comment (ZoomRmc) [inline searches/binary_search.nim:72]:
```suggestion
    check binarySearchRecursive(arr, key) == expected
```
*If the overload above is accepted.

**Review by ZoomRmc — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please, consider the proposed changes.

File: searches/binary_search.nim (hunk 1 above)
Comment (dlesnoff) [reply searches/binary_search.nim:101]: That's a suggestion I should have made at the Linear Search PR.

File: searches/binary_search.nim [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      return some(Natural(mid))
+
+    if key < arr[mid]: 
+      right = mid - 1
+    else:
+      left = mid + 1
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  none(Natural)
+
+func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
```
Comment (dlesnoff) [inline searches/binary_search.nim:53]:
```suggestion
func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: Natural, key: T): Option[Natural] =
```

File: searches/binary_search.nim [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+## =============
+## Binary search is an efficient searching algorithm. It is faster than linear 
+## search except for small arrays. However, binary search can only be used on sorted arrays.
+## If the array is not sorted, it must be sorted first before binary search can be applied.
+##
+## Binary search starts by comparing the key value with the middle element of the array.
+## If the key value matches the middle element, the search is complete.
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
```
Comment (dlesnoff) [inline searches/binary_search.nim:11]: Add a *small* note on other approaches like linear search and search in a tree. See: https://en.wikipedia.org/wiki/Binary_search_algorithm#Binary_search_versus_other_schemes

File: searches/binary_search.nim [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    if key < arr[mid]: 
+      right = mid - 1
+    else:
+      left = mid + 1
+  # `none(Natural)` is returned when both halfs are empty after some iterations.
+  none(Natural)
+
+func binarySearchRecursive[T:Ordinal](arr: openArray[T], left, right: int, key: T): Option[Natural] =
+  if left <= right:
+    let mid = left + (right - left) div 2
```
[the remark this answers is not in the document]
Comment (dlesnoff) [reply searches/binary_search.nim:55]: Forget this if you made the change above

File: searches/binary_search.nim (hunk 5 above)
Comment (Archargelod) [reply searches/binary_search.nim:66]: Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). Negative indices can be easily handled by the algorithm without complex edge cases.

Comment (ZoomRmc) [reply searches/binary_search.nim:66]: > Natural here really only adds the complexity as `right` could be -1 (when `mid` = 0). I believe, negative indices can be easily handled by the algorithm without complex edge cases.

There's no negative indexes for `openArrays` and they are necessary to be array/seq agnostic. The other options is to use a concept that will embrace ordinal-indexed types or write type-specific overloads.

File: searches/binary_search.nim [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+## Other common choices for pivot include the first or last element of the array,
+## a random element, or the median of three values (the first, middle, and last elements).
+##
+## Note, it is common for many binary search algorithms to include options for
+## finding the right-most or left-most occurrence of the key element.
+## However, this implementation does not include these options.
+## It is intended for simple searches where the presence of the key element is all that matters.
+## https://en.wikipedia.org/wiki/Binary_search_algorithm#Duplicate_elements
+##
+## If the input is not already sorted, it may be faster to use a linear search instead.
```
Comment (dlesnoff) [inline searches/binary_search.nim:32]: ChatGPT is not confident. Let us be. Here we can.
```suggestion
## If the input is not already sorted, it is faster to use a linear search instead.
```

**Review by dlesnoff — CHANGES_REQUESTED** [review]
I changed my mind. Let us just refer to Wikipedia for precisions on theoretical aspects, it will ever be more reviewed and better referenced as well.
Simplify the docstring as suggested, and I'll finally accept the changes.

File: searches/binary_search.nim [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+## Binary search is an efficient algorithm for searching sorted arrays,
+## generally outperforming linear search except for small arrays. 
+## However, binary search requires the array to be pre-sorted.
+##
+## Binary search starts by comparing the key value with the middle element of the array.
+## If the key value matches the middle element, the search is complete.
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
+## https://en.wikipedia.org/wiki/Binary_search_algorithm
```
Comment (dlesnoff) [inline searches/binary_search.nim:12]: Move this reference to the end of the docstring.

File: searches/binary_search.nim [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+## Average and Worst Time Complexity: O(log n), where n is the length of the array.
+## Space Complexity in iterative approach: O(1)
+## Space Complexity in recursive approach: O(n)
+##
+## The choice of pivot can greatly affect the performance of a binary search algorithm.
+## This implementation uses the middle element of the array as pivot, it works well
+## when the array is evenly distributed and balanced, as it allows
+## for cutting the search in half with each iteration.
+## Other common choices for pivot include the first or last element of the array,
+## a random element, or the median of three values (the first, middle, and last elements).
```
[the remark this answers is not in the document]
Comment (Archargelod) [reply searches/binary_search.nim:24]: > Is this A.I. generated?
Yes, sorry for that. I'm not that fluent in english, so I cheated just a bit with docs. 
But don't worry, as code is 100% handwritten and not generated.

[the remark this answers is not in the document]
Comment (ZoomRmc) [reply searches/binary_search.nim:24]: Next time just say "I could use some help with the docs". Someone will surely offer some necessary additions.

[the remark this answers is not in the document]
Comment (dlesnoff) [reply searches/binary_search.nim:24]: > But don't worry, as code is 100% handwritten and not generated.

I trust you on that. It is not so much that it has been generated by A.I. but that the documentation was *wrong*.
The code quality is good.

**Comment by dlesnoff** [comment]
Run nimpretty to remove trailing whitespaces. You can wrap the arguments on one line by adding a line break after the parenthesis instead of the automatically inserted line break by nimpretty.
I will merge it immediately after.

**Review by dlesnoff — APPROVED** [review]
Nice first contribution!

**Pull request merged**